numpy/core/src/multiarray/buffer.c

```diff
@@ -771,17 +771,6 @@ array_getbuffer(PyObject *obj, Py_buffer *view, int flags)
             goto fail;
         }
     }
-    /*
-     * If a read-only buffer is requested on a read-write array, we return a
-     * read-write buffer, which is dubious behavior. But that's why this call
-     * is guarded by PyArray_ISWRITEABLE rather than (flags &
-     * PyBUF_WRITEABLE).
-     */
-    if (PyArray_ISWRITEABLE(self)) {
-        if (array_might_be_written(self) < 0) {
-            goto fail;
-        }
-    }
```

Comment:
As an alternative to speeding this deprecation, is it sufficient to change this and the change below to:
```C
if (!PyArray_CHKFLAGS(self, NPY_ARRAY_WARN_ON_WRITE)) {
    readonly = PyArray_ISWRITEABLE(self);  // keep the old weird behavior and comment above
}
else if ((flags & PyBUF_WRITEABLE) == PyBUF_WRITEABLE) {
    // emit the warning from array_might_be_written
    readonly = False;
}
else {
    // return a readonly-array since doing so skirts the need for a warning
    readonly = True
}
```

Reply:
I think I misunderstood the below, what I suggest is equivalent to what you already have and probably not any clearer.